src/compiler/parser.ts

```diff
@@ -1940,7 +1940,7 @@ namespace ts {
             return finishNode(node);
         }
 
-        function parseTypePredicate(lhs: Identifier | ThisTypeNode): TypePredicateNode {
+        function parseThisTypePredicate(lhs: Identifier | ThisTypeNode): TypePredicateNode {

```

Comment:
If this only parses `ThisTypePredicate`s, then it should probably only accept a `ThisTypeNode` as an argument. (Since `Identifier` would always imply an Identifier-based type predicate)